channeldb/mp_payment.go

```diff
@@ -0,0 +1,109 @@
+)
+
+// MPPaymentCreationInfo is the information necessary to have ready when
+// initiating a payment, moving it into state InFlight.
+type MPPaymentCreationInfo struct {
+	// PaymentHash is the hash this payment is paying to.
+	PaymentHash lntypes.Hash
+
+	// PaymentAddr is a receiver-generated nonce used to prevent concurrent
+	// payments from colliding.
+	PaymentAddr lntypes.Hash
+
+	// Value is the amount we are paying.
+	Value lnwire.MilliSatoshi
+
+	// CreatingTime is the time at which this payment was started.
+	CreationTime time.Time
+
+	// PaymentRequest is the full payment request, if any.
+	PaymentRequest []byte
+}
+
+// HTLCAttempt contains information about a specific HTLC attempt for a given
+// payment. This information is used by the router to handle any errors coming
+// back after an attempt is made, and to query the switch about the status of a
+// payment. For settled payment this will be the information for the succeeding
+// payment attempt.
+type HTLCAttempt struct {
+	// PaymentID is the unique ID used for this attempt.
+	PaymentID uint64
+
+	// SessionKey is the ephemeral key used for this payment attempt.
+	SessionKey *btcec.PrivateKey
+
+	// Route is the route attempted to send the HTLC.
+	Route route.Route
+
+	// AttemptTime is the time at which this HTLC was attempted.
+	AttemptTime time.Time
+
+	// Settle is the preimage of a successful payment. This serves as a
+	// proof of payment. It will only be non-nil for settled payments.
+	//
+	// NOTE: Can be nil if payment is not settled.
+	Settle *HTLCSettleInfo
+
+	// Fail is a failure reason code indicating the reason the payment
+	// failed. It is only non-nil for failed payments.
+	//
+	// NOTE: Can be nil if payment is not failed.
+	Failure *HTLCFailInfo
+}
+
+// HTLCSettleInfo encapsulates the information that augments an HTLCAttempt in
+// the event that the HTLC is successful.
+type HTLCSettleInfo struct {
+	// Preimage is the preimage of a successful HTLC. This serves as a proof
+	// of payment.
+	Preimage lntypes.Preimage
+
+	// SettleTime is the time at which this HTLC was settled.
+	SettleTime time.Time
+}
+
+// HTLCFailInfo encapsulates the information that augments an HTLCAttempt in the
+// event that the HTLC fails.
+type HTLCFailInfo struct {
+	// FailTime is the time at which this HTLC was failed.
+	FailTime time.Time
+}
+
+// MPPayment is a wrapper around a payment's MPPaymentCreationInfo and
+// HTLCAttempts. All payments will have the MPPPaymentCreationInfo set, any
+// HTLCs made in attempts to be completed will populated in the HTLCs slice.
+// Each populated HTLCAttempt represents an attempted HTLC, each of which may
+// have the associated Settle or Fail struct populated if the HTLC is no longer
+// in-flight.
+type MPPayment struct {
+	// sequenceNum is a unique identifier used to sort the payments in
+	// order of creation.
+	sequenceNum uint64
+
+	// Status is the current PaymentStatus of this payment.
+	Status PaymentStatus
```

Comment:
This `PaymentStatus` is mostly a product of the raw outcomes of the attempted htlcs for this payment. I am not sure how to deal with that. `FailureReasonTimeout` is not related to any of the htlcs, because it is a failure at the level of the payment loop itself. `FailureReasonError` can be several things. 

Do we also need to store a different type of state on the htlc level? If a new htlc is added for an mp payment, we need to rederive the top payment-level status. I am not sure, just feel that there is something here.

I don't mean reverting the change you just made. I am sure we need a payment-level status. But it may be useful to think through some multi-path scenarios and see how we'd update that payment-level status.

Reply:
yes at the moment this just moves the existing field to the top-level payment, and since we only have one htlc there shouldn't be any gaps.

as we move to real MPP, there are some additional states we would need to consider. do you think they should be addressed in this PR?

> we need to rederive the top payment-level status.

Yes my plan is that this field will be populated as a pure function of the htlc attempts and their individual states

Reply:
No, doesn't need to be in this PR. But won't there by a migration problem when you will do that pure function, but don't have the htlc outcomes?

Reply:
Currently we assume the payment is being actively tried if settle/fail is not present, and there's an associated attempt known by the switch. I think you are right that this will be a bit different, as there are both HTLC-level failures and payment level failures possible.  I don't think it needs to be it's own field though, as we can derive it from the HTLCs and FailureReason.

Reply:
Discussed offline. If we want to make `SendToRoute` reliable too (left over from the reliable payment project), we need a way to retrieve the raw htlc failure with `TrackPayment`. It seems that we need to add it before we do the great payment db migration